presto-verifier/src/main/java/com/facebook/presto/verifier/framework/LimitQueryDeterminismAnalyzer.java

```diff
@@ -0,0 +1,143 @@
+
+import java.util.Optional;
+
+import static com.facebook.presto.sql.QueryUtil.simpleQuery;
+import static com.facebook.presto.verifier.framework.LimitQueryDeterminismAnalyzer.Analysis.DETERMINISTIC;
+import static com.facebook.presto.verifier.framework.LimitQueryDeterminismAnalyzer.Analysis.FAILED_DATA_CHANGED;
+import static com.facebook.presto.verifier.framework.LimitQueryDeterminismAnalyzer.Analysis.NON_DETERMINISTIC;
+import static com.facebook.presto.verifier.framework.LimitQueryDeterminismAnalyzer.Analysis.NOT_RUN;
+import static com.facebook.presto.verifier.framework.QueryStage.DETERMINISM_ANALYSIS;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.Objects.requireNonNull;
+
+public class LimitQueryDeterminismAnalyzer
+{
+    public enum Analysis
+    {
+        NOT_RUN,
+        NON_DETERMINISTIC,
+        DETERMINISTIC,
+        FAILED_DATA_CHANGED,
+    }
+
+    private final PrestoAction prestoAction;
+    private final boolean enabled;
+
+    public LimitQueryDeterminismAnalyzer(PrestoAction prestoAction, VerifierConfig verifierConfig)
+    {
+        this.prestoAction = requireNonNull(prestoAction, "prestoAction is null");
+        this.enabled = verifierConfig.isEnableLimitQueryDeterminismAnalyzer();
+    }
+
+    public Analysis analyze(QueryBundle control, long rowCount)
+    {
+        if (!enabled) {
+            return NOT_RUN;
+        }
+
+        Statement statement = control.getQuery();
+        Query query;
+
+        // A query is rewritten to either an Insert or a CreateTableAsSelect
+        if (statement instanceof Insert) {
+            query = ((Insert) statement).getQuery();
+        }
+        else if (statement instanceof CreateTableAsSelect) {
+            query = ((CreateTableAsSelect) statement).getQuery();
+        }
+        else {
+            return NOT_RUN;
+        }
+
+        // Flatten TableSubquery
+        if (query.getQueryBody() instanceof TableSubquery) {
+            Optional<With> with = query.getWith();
+            while (query.getQueryBody() instanceof TableSubquery) {
+                // ORDER BY and LIMIT must be empty according to syntax
+                if (query.getOrderBy().isPresent() || query.getLimit().isPresent()) {
+                    return NOT_RUN;
+                }
+                query = ((TableSubquery) query.getQueryBody()).getQuery();
+                // WITH must be empty according to syntax
+                if (query.getWith().isPresent()) {
+                    return NOT_RUN;
+                }
+            }
+            query = new Query(with, query.getQueryBody(), query.getOrderBy(), query.getLimit());
+        }
+
+        // If Query contains ORDER_BY, we won't be able to make conclusion even if there is LIMIT clause.
+        if (query.getOrderBy().isPresent()) {
+            return NOT_RUN;
+        }
+
+        Query queryNoLimit;
+        if (query.getLimit().isPresent()) {
```

Comment:
It will be false.
`Query.getLimit().isPresent()` is true iff the select query has a LIMIT clause on the top level and the query is a set operation: `union`, `intersect`, `except`.
`QuerySpecification.getLimit().isPresent()` is true iff the select query has a LIMIT clause on the top level, and the query is not a set operation.